### PR TITLE
Factor topology into package: PlutoDependencyExplorer.jl

### PR DESCRIPTION
**Pull request by fonsp**
We already did #2698 which means that people can find the references and definitions of an `Expr` in another package more easily. This PR factors out a second package that layers on top of ExpressionExplorer: given a list of `Expr` (i.e. a notebook of cells), in which order should the expressions be evaluated?

Not sure about the name PlutoReactiveCore

TODO:
- [x] Factor out the `Cell`. PlutoReactiveCore only uses it as a key (like `Dict{Cell,ReactiveNode}`), without accessing any of its fields. (This was the result of previous refactors.) To solve the dependency, it could be replaced with a `UUID` everywhere. But this will means lots of conversions everywhere in our code that's less charming. We could also use an `AbstractCell` type in PlutoReactiveCore, with a `Cell <: AbstractCell` in Pluto.
- [x] Factor our the `PlutoRuner.@bind` dependency used in the static macro expand of ExpressionExplorerExtras. Easiest is probably to write a copy of the macro inside PlutoReactiveCore, since it's only used for analysis.
- [x] Make sure that it works inside the Pluto codebase as a `module`, and put into a separate package afterwards.

### CODE REVIEW

**Comment by fonsp** [comment]
@quildtide you would like this!

**Comment by fonsp** [comment]
@pankgeorg any name suggestions?

File: src/analysis/PlutoReactiveCore/ExpressionExplorer.jl [hunk 1]
```diff
@@ -78,17 +91,39 @@ function macro_has_special_heuristic_inside(; symstate::SymbolsState, expr::Expr
...
+const found_plutorunner = Ref{Union{Nothing,Module}}(nothing)
+
+function get_plutorunner()
+    fpr = found_plutorunner[]
+    if fpr === nothing
+        # lets try really hard to find it!
+        if haskey(Base.loaded_modules, plutorunner_id)
+            found_plutorunner[] = Base.loaded_modules[plutorunner_id]
+        elseif haskey(Base.loaded_modules, pluto_id)
+            found_plutorunner[] = Base.loaded_modules[pluto_id].PlutoRunner
```
Comment (Pangoraw) [inline src/analysis/PlutoReactiveCore/ExpressionExplorer.jl:116]: I don't think we want this one first. If Pluto is loaded in the process then Pluto.PlutoRunner should be preferred (supposing that PlutoReactiveCore is most likely being called by Pluto).
```suggestion
        if haskey(Base.loaded_modules, pluto_id)
            found_plutorunner[] = Base.loaded_modules[pluto_id].PlutoRunner
        elseif haskey(Base.loaded_modules, plutorunner_id)
            found_plutorunner[] = Base.loaded_modules[plutorunner_id]
```

**Review by Pangoraw — COMMENTED** [review]
Nice! I think the added complexity is justifiable if we can use it directly in other packages such as PlutoUI for `@div` or maybe we can get away with a simpler version in PlutoUI without the extra complexity?

**Comment by pankgeorg** [comment]
> Not sure about the name PlutoReactiveCore

It feels that this isn't the reactive core; the reactive core _uses_ the topology that this package outlines IIUC.

I'd call this `ReactivityExplorer` or `DependencyExplorer` or `OrderExplorer`, following `ExpressionExplorer` or something

**Comment by fonsp** [comment]
yolooo

**Pull request merged**